### DISCUSSION TITLE
nonce empty on client

### DISCUSSION THREAD

**Discussion by curler7** [epicweb-dev/epic-stack#768] [Q&A]
## Getting following error on browser console
Warning: Prop `nonce` did not match. Server: "0d1781b142423cac737af4b3e38a32ea" Client: ""

### Getting started with the following instructions
git clone https://github.com/epicweb-dev/epic-stack.git
cd epic-stack
npm install
cp .env.example .env
npm run setup

Tested with firefox and chromium. Both displayed same error.
Did i missed something? Thx for response.

**Comment by kentcdodds** [comment]
Hi @curler7,

I can't reproduce this on the deployed version of the stack: https://www.epicstack.dev

**Reply by kennethvdberghe** [reply]
I see that all nonce values are empty on de deployed version though
![image](https://github.com/epicweb-dev/epic-stack/assets/68184728/bfe4ccbe-749e-47e2-a581-1d2247854428)

**Reply by huangc28** [reply]
I have the same problem. The client nonce is an empty value, but nonce generated on entry.server.tsx has value. This causes my web to have difference render result between the server and the client.

**Reply by kiliman** [reply]
Chrome DevTools doesn't display the `nonce` in the Elements tab. You can see that the `nonce` value is set if you View Source.

Also, you can see the `nonce` values by entering the following in the DevTools console:
```js
$$('script').map(s => s.nonce)
```
<img width="1325" alt="image" src="https://github.com/user-attachments/assets/77b84cc3-f116-4636-9aeb-c865a90efe75">

If you're getting hydration warnings, it's probably due to a browser extension instead of the `nonce` values. You can verify by opening your web app in an Incognito window.

To prevent hydration warnings, I would recommend updating react and react-dom to latest RC versions. React 19 fixes the warnings caused by full-document hydration.

**Comment by reify-davidhale** [comment]
Adding my $0.02 to the discussion. 

When I run [Firefox Developer Edition](https://www.mozilla.org/en-US/firefox/developer/) and these initial commands: 
```sh
npx create-epic-app@latest

# didn't setup deploy. picked random name. 
cd <randomly_picked_name> 
npm run dev
```

Upon opening the app I received the following screenshot: (validated while Private Browsing) 
<img width="1750" alt="Screenshot 2024-08-26 at 2 52 38 PM" src="https://github.com/user-attachments/assets/4d29de7f-d5de-4d53-9036-da578a142ec7">

Checked in private browsing as recommended by @kiliman and still receiving this. To add further note, I am not seeing the same thing in Chrome DevTools locally. Surprisingly enough even though the `nonce` value is considered `""` to begin with checking the Inspector yields `nonce` values which are not empty 🤔 Also once I deploy a production build for the `react-router-dom` entry the hydration issue seems to rectify as well. 

<img width="439" alt="Screenshot 2024-08-26 at 2 55 44 PM" src="https://github.com/user-attachments/assets/a931d18f-2294-47d8-825e-a8317636f19c">

> To prevent hydration warnings, I would recommend updating react and react-dom to latest RC versions. React 19 fixes the warnings caused by full-document hydration.

I assume this is the proper solution is just await React 19? ☝️ 

Let me know and thanks for reading!

**Reply by kentcdodds** [reply]
I am able to reproduce this only when running the dev server and only with firefox (also tested chromium-based browsers and safari). If I run the production build locally (or deployed) I don't get an error in any browser.

This is odd, surprising, and annoying, but will not affect the production application. Considering it's only affecting Firefox, I'm going to have to assume this is a firefox bug. Sorry I can't be more helpful!

**Reply by reify-davidhale** [reply]
No this is great @kentcdodds! I really thought I was taking crazy pills so I now know I was only half-crazy :)

**Comment by lhapaipai** [comment]
Hi,
Thanks for this amazing repo, source of inspiration.
I want to reopen the discussion.
Indeed this issue is only highlighted in Firefox but I think it is present.

the `NonceProvider` is only defined server side

```tsx
// entry.server.tsx

export default async function handleRequest(...args: DocRequestArgs) {
	//...

	const nonce = crypto.randomBytes(16).toString('hex')
	return new Promise(async (resolve, reject) => {
                //...
		const { pipe, abort } = renderToPipeableStream(
			<NonceProvider value={nonce}>
				<ServerRouter
					nonce={nonce}
					context={reactRouterContext}
					url={request.url}
				/>
			</NonceProvider>,
                        {
                          //...
                          nonce
                        }
		)

		setTimeout(abort, streamTimeout + 5000)
	})
}
```
but not client Side
```tsx
// entry.client.tsx
startTransition(() => {
	hydrateRoot(document, <HydratedRouter />)
})
```
```tsx
// root.tsx
export function Layout({ children }: { children: React.ReactNode }) {
	// if there was an error running the loader, data could be missing
	const data = useLoaderData<typeof loader | null>()
	
        const nonce = useNonce()
        // correct value server side
        // default value of NonceContext client side -> empty string
}
```
to highlight the issue.
define a default value of NonceContext.
```ts
// nonce-provider.ts
export const NonceContext = React.createContext<string>('hello');
```
we have a hydration error on all browsers

with Firefox
```diff
    <script
+              nonce="hello" // default value client side
-              nonce="edd3ae271e08d972ebd0d3440a1b1909" // server side
      dangerouslySetInnerHTML={{__html:"\n// This ..."}}
    >
```

with Chrome based

```diff
    <script
+              nonce="hello" // default value client side
-              nonce=""      // not accessible ?? but present in the source code
      dangerouslySetInnerHTML={{__html:"\n// This ..."}}
    >
```


What do you think?

**Reply by kentcdodds** [reply]
Chrome intentionally hides the nonce for security reasons.

**Comment by lhapaipai** [comment]
I'm stupid, the goal being precisely not to be able to access it on the client side...
So finally the entry point scripts have a nonce and the 'strict-dynamic' rule allows you not to worry about the scripts that will be loaded on the client side by the scripts defined on the server side. Is that right?
Sorry, I'm asking questions...

However, I have an answer about Firefox's behavior. If you set `reportOnly` to false, it behaves like Chrome.
